### ISSUE TITLE
HTTP 405 should be HTTP 501

### ISSUE THREAD

**Issue by alexykot** [pypiserver/pypiserver#165]
On the endpoints that are not implemented (basically everything apart from `upload` and `install`) server currently returns HTTP 405. It should return HTTP 501 as that would be correct way to handle valid API requests that are not yet supported by server.

**Comment by mplanchard** [comment]
Thanks for pointing this out. I definitely agree that 405 ("method not allowed") is not the proper status code. However, I'm also not sure that 501 is the correct code. The official 501 definition is (via w3.org):

> The server does not support the functionality required to fulfill the request. This is the appropriate response when the server does not recognize the request method and is not capable of supporting it for any resource. 

I'd say it's more accurately a 404:

> The server has not found anything matching the Request-URI. No indication is given of whether the condition is temporary or permanent. The 410 (Gone) status code SHOULD be used if the server knows, through some internally configurable mechanism, that an old resource is permanently unavailable and has no forwarding address. This status code is commonly used when the server does not wish to reveal exactly why the request has been refused, or when no other response is applicable. 

And in fact, most web frameworks (e.g. Flask, Django) will return a 404 if an endpoint is requested that has not been configured. What are your feelings on switching from 405 to 404?

**Comment by ajd394** [comment]
I agree with @alexykot. The server should only respond with 400s when a client had issued a request that is incorrect. Returning 501 will make it clear to users that this server doesn't support the requested pypi functionality.

**Comment by mplanchard** [comment]
So how about  mapping out the end points on Pypi and returning 501 for
those, and 404 for others? I would prefer not to return NotImplemented for
every possible nonexistent endpoint, but am fine with doing so for a
limited set that people might expect.

@alexyko or @ajd394, would either of you be able to provide a listing of
endpoints for which we should return a 501 rather than a 404?

**Comment by ajd394** [comment]
@mplanchard I agree that it should only return 501 for non implemented pypi endpoints. I have been poking around the PEP to find a document that specifies all of the pypi endpoints but i haven't found it yet.

**Comment by mplanchard** [comment]
Let us know if you find something definite. If not, we can start by compiling a list of all the endpoints of which we're aware and adjust going forward from there.

**Comment by mplanchard** [comment]
@alexykot @ajd394 [these](https://github.com/pypa/warehouse/blob/master/warehouse/routes.py) are the endpoints implemented by `Warehouse`, which is the "new" PyPI server. It's harder to find something for "legacy-pypi." Any feelings about that list of endpoints as being one for which we should handle with 501's if we don't implement them?
